### PR TITLE
[Superseded] [Quark] Cache dequantized MXFP4 fallback weights

### PR DESCRIPTION
**Pull request by LiuYinfeng01**
> **Superseded by #50814. This PR is retained only for history.**

## Summary

- Add `VLLM_QUARK_MX_CACHE_DEQUANT_WEIGHT=1` as an explicit opt-in for Quark MXFP4 linear emulation.
- Dequantize packed MXFP4 weights once after loading instead of on every linear invocation.
- Preserve activation quantize/dequantize semantics.

## Motivation

The emulation backend normally performs the same packed-weight dequantization for every invocation. For decode workloads this repeated work can dominate the shared/dense linear path. Caching moves that work to model load time, trading device memory for lower inference latency.

## Technical approach

The optimization activates only when:

1. `VLLM_QUARK_MX_CACHE_DEQUANT_WEIGHT=1` is set; and
2. the selected backend is `EmulationMxfp4LinearKernel`.

At load time:

- dynamic MXFP4 weight quantization still runs first when applicable;
- the packed E2M1 weight and E8M0 scale are dequantized using the model parameter dtype;
- the result is installed as a non-trainable parameter on the original device;
- `weight_scale` is unregistered after it is no longer needed;
- a repeated processing call is a no-op, so dequantization occurs only once.

At apply time, the cached weight is passed to `F.linear`, but the emulation backend's activation QDQ function is still called for every invocation.

## Precision and semantic boundary

This PR changes only the lifetime of weight dequantization. It intentionally does not implement or enable `VLLM_QUARK_MX_SKIP_ACTIVATION_QDQ`.

Therefore:

- packed checkpoint weights are unchanged on disk;
- activation QDQ remains part of the numerical path;
- native MXFP4 backends are unchanged;
- switch-off behavior remains the existing per-call dequantization path.

## Tests

```bash
python3 -m pytest -q tests/quantization/test_quark_ocp_mx_cache.py

ruff check \
  vllm/model_executor/layers/quantization/quark/schemes/quark_ocp_mx.py \
  tests/quantization/test_quark_ocp_mx_cache.py

ruff format --check \
  vllm/model_executor/layers/quantization/quark/schemes/quark_ocp_mx.py \
  tests/quantization/test_quark_ocp_mx_cache.py
```

Coverage and results:

- switch off preserves the original kernel path;
- switch on dequantizes once;
- `weight_scale` is released;
- cached and uncached outputs are close;
- cached dtype/device and `requires_grad=False` are preserved;
- activation QDQ is retained;
- total: `5 passed`;
- Ruff check and format check: passed.

## Historical performance evidence

The following older-runtime five-tier measurements used the cached fallback:

- con1: `90.97 ms` mean TPOT, `10.11 tok/s`;
- con4: `109.98 ms`, `31.59 tok/s`;
- con16: `153.79 ms`, `68.71 tok/s`;
- con32: `273.65 ms`, `86.36 tok/s`;
- con128: `288.38 ms`, `88.82 tok/s`.

On that runtime, con1 improved from full emulation `402.43 ms` to `90.97 ms`. The previously recorded con4 comparison was `177.14 ms -> 109.98 ms`.

On the newer #4450 attention runtime, the combined routed-A8W4 plus cached shared/dense fallback measured:

- con1 mean TPOT: `22.551 ms`;
- output throughput: `33.123 tok/s`;
- successful requests: `3/3`;
- versus BF16 `22.995 ms`: `-1.93%` TPOT.

The old-runtime values and #4450 con1 are different runtime populations and are not presented as one strict A/B series.

## Historical GSM8K evidence

For the K3 integration configuration:

- official emulation: `96.36% flexible`, `96.36% strict`;
- routed A8W4 plus per-call dense emulation: `96.21% flexible`, `96.21% strict`;
- routed A8W4 plus cached dense fallback: `96.29% flexible`, `96.36% strict`.

The cached configuration changed accuracy by `-0.07/+0.00 percentage points` versus official emulation and stayed within the `<=0.5 percentage point` gate.

These are model-level integration results, not generic unit-test guarantees. They also must not be attributed to the separate native A4W4 path, whose complete GSM8K 8-shot run is still pending.

## Memory trade-off

Observed on the K3 workload:

- approximately `+1.5 GiB` resident dequantized-weight memory;
- approximately `3.43 GiB` less memory available for KV cache.

The exact increase depends on model dimensions, dtype, and the number of emulated layers.

## Risks

- Extra resident weight memory can reduce KV-cache capacity or cause OOM.
- The environment switch is currently read directly rather than registered in the central vLLM environment registry.
- State-dict and distributed lifecycle behavior should receive broader integration coverage beyond the focused unit tests.

## Reviewer notes

Please review the parameter lifecycle, idempotence, activation-QDQ preservation, and whether the environment switch should move into the central registry. This first change intentionally avoids the separate skip-activation-QDQ semantic change.

### CODE REVIEW

**Comment by LiuYinfeng01** [comment]
Superseded by #50814, which contains the identical change as a DCO-signed single commit.
